x-pack/plugins/infra/server/lib/log_analysis/log_entry_anomalies.ts

```diff
@@ -0,0 +1,391 @@
+    typical,
+    actual,
+    duration,
+    startTime: anomalyStartTime,
+    categoryId,
+    type: 'logCategory' as const,
+  };
+};
+
+async function fetchLogEntryAnomalies(
+  mlSystem: MlSystem,
+  jobIds: string[],
+  startTime: number,
+  endTime: number,
+  sort: Sort,
+  pagination: Pagination
+) {
+  // We'll request 1 extra entry on top of our pageSize to determine if there are
+  // more entries to be fetched. This avoids scenarios where the client side can't
+  // determine if entries.length === pageSize actually means there are more entries / next page
+  // or not.
+  const expandedPagination = { ...pagination, pageSize: pagination.pageSize + 1 };
+
+  const finalizeFetchLogEntryAnomaliesSpan = startTracingSpan('fetch log entry anomalies');
+
+  const results = decodeOrThrow(logEntryAnomaliesResponseRT)(
+    await mlSystem.mlAnomalySearch(
+      createLogEntryAnomaliesQuery(jobIds, startTime, endTime, sort, expandedPagination)
+    )
+  );
+
+  const {
+    hits: { hits },
+  } = results;
+  const hasMoreEntries = hits.length > pagination.pageSize;
+
+  // An extra entry was found and hasMoreEntries has been determined, the extra entry can be removed.
+  if (hasMoreEntries) {
+    hits.pop();
+  }
+
+  // To "search_before" the sort order will have been reversed for ES.
+  // The results are now reversed back, to match the requested sort.
+  if (pagination.cursor && 'searchBefore' in pagination.cursor) {
+    hits.reverse();
+  }
+
+  const paginationCursors =
+    hits.length > 0
+      ? {
+          previousPageCursor: hits[0].sort,
+          nextPageCursor: hits[hits.length - 1].sort,
+        }
+      : undefined;
+
+  const anomalies = hits.map((result) => {
+    const {
+      job_id,
+      record_score: anomalyScore,
+      typical,
+      actual,
+      partition_field_value: dataset,
+      bucket_span: duration,
+      timestamp: anomalyStartTime,
+      by_field_value: categoryId,
+    } = result._source;
+
+    return {
+      id: result._id,
+      anomalyScore,
+      dataset,
+      typical: typical[0],
+      actual: actual[0],
+      jobId: job_id,
+      startTime: anomalyStartTime,
+      duration: duration * 1000,
+      categoryId,
+    };
+  });
+
+  const fetchLogEntryAnomaliesSpan = finalizeFetchLogEntryAnomaliesSpan();
+
+  return {
+    anomalies,
+    paginationCursors,
+    hasMoreEntries,
+    timing: {
+      spans: [fetchLogEntryAnomaliesSpan],
+    },
+  };
+}
+
+export async function getLogEntryExamples(
+  context: RequestHandlerContext & { infra: Required<InfraRequestHandlerContext> },
+  sourceId: string,
+  startTime: number,
+  endTime: number,
+  dataset: string,
+  exampleCount: number,
+  sourceConfiguration: InfraSource,
+  callWithRequest: KibanaFramework['callWithRequest'],
+  categoryId?: string
+) {
+  const finalizeLogEntryExamplesSpan = startTracingSpan('get log entry rate example log entries');
+
+  const jobId = getJobId(context.infra.spaceId, sourceId, 'log-entry-rate');
```

Comment:
We would probably run into problems anyway if the log rate and category jobs were configured with different timestamp fields.

Here, though, it shouldn't be too hard to at least try to make the distinction between the two jobs. Wouldn't the presence of the `categoryId` be a suitable criterion to switch between the job types?

```suggestion
  const jobId = getJobId(
    context.infra.spaceId,
    sourceId,
    categoryId != null ? logEntryCategoriesJobTypes[0] : logEntryRateJobTypes[0]
  );
```

Reply:
Absolutely, this should have been handled 👍 `getLogEntryExamples` was originally only handling the one job type from the original PR, and I just forgot to port over the differentation here. Didn't notice it when I was layering on the `categoryId` (for the pattern) changes.